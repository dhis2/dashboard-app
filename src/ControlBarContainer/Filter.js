import React, { Component, Fragment } from 'react';
import PropTypes from 'prop-types';
import { grey700 } from 'material-ui/styles/colors';
import TextField from 'material-ui/TextField';
import IconButton from 'material-ui/IconButton';
import IconClear from 'material-ui/svg-icons/content/clear';
import isEmpty from 'd2-utilizr/lib/isEmpty';

import * as fromReducers from '../reducers';

const KEYCODE_ESCAPE = 27;

const styles = {
    filterField: {
        fontSize: '14px',
        width: '200px',
        height: '30px',
    },
    filterFieldInput: {
        top: '-9px',
        left: '1px',
    },
    filterFieldUnderline: {
        bottom: '10px',
    },
    filterFieldUnderlineFocus: {
        borderColor: '#aaa',
        borderWidth: '1px',
    },
    clearButton: {
        width: '20px',
        height: '24px',
        padding: 0,
        top: '-5px',
        left: '-18px',
    },
    clearButtonIcon: {
        width: '16px',
        height: '16px',
    },
};

class Filter extends Component {
    constructor(props) {
        super(props);

        this.state = {
            value: fromReducers.fromFilter.DEFAULT_NAME,
        };

        this.setFilterValue = this.setFilterValue.bind(this);
        this.handleKeyUp = this.handleKeyUp.bind(this);
    }

    componentWillReceiveProps(nextProps) {
        this.setState({
            value: nextProps.name,
        });
    }

    setFilterValue(event) {
        event.preventDefault();

        this.props.onChangeName(event.target.value);
    }

    handleKeyUp(event) {
        if (event.keyCode === KEYCODE_ESCAPE) {
            this.props.onChangeName();
        }
    }

    render() {
        return (
            <TextField
                className="FilterField"
                value={this.state.value}
                onChange={this.setFilterValue}
                onKeyUp={this.handleKeyUp}
                hintText="Filter dashboards"
                style={styles.filterField}
                inputStyle={styles.filterFieldInput}
                hintStyle={styles.filterFieldHint}
                underlineStyle={styles.filterFieldUnderline}
                underlineFocusStyle={styles.filterFieldUnderlineFocus}
            />
        );
    }
}

Filter.propTypes = {
    name: PropTypes.string,
    onChangeName: PropTypes.func,
};

Filter.defaultProps = {
    name: '',
    onChangeName: Function.prototype,
};

const ClearButton = ({ name, onChangeName }) => {
    const disabled = isEmpty(name);

    return (
        <IconButton
            style={Object.assign({}, styles.clearButton, {
                opacity: disabled ? 0 : 1,
            })}
            iconStyle={styles.clearButtonIcon}
            onClick={() => onChangeName()}
            disabled={disabled}
        >
            <IconClear color={grey700} />
        </IconButton>
    );
};

ClearButton.propTypes = {
    name: PropTypes.string.isRequired,
    onChangeName: PropTypes.func.isRequired,
};

export default props => (
    <Fragment>
        <Filter {...props} />
        <ClearButton {...props} />
    </Fragment>
);
