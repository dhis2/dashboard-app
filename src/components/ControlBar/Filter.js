import React, { Component } from 'react';
import PropTypes from 'prop-types';
import { withStyles } from '@material-ui/core/styles';
import i18n from '@dhis2/d2-i18n';
import InputField from '@material-ui/core/Input';
import InputAdornment from '@material-ui/core/InputAdornment';
import SearchIcon from '@material-ui/icons/Search';
import { colors } from '@dhis2/ui-core';

import ClearButton from './ClearButton';
import { DEFAULT_STATE_DASHBOARDS_FILTER_NAME } from '../../reducers/dashboardsFilter';

export const KEYCODE_ENTER = 13;
export const KEYCODE_ESCAPE = 27;

const styles = {
    filterField: {
        fontSize: '14px',
        width: '200px',
        height: '30px',
        top: '-4px',
    },
    searchIcon: {
        color: colors.grey700,
        width: '20px',
        height: '20px',
    },
};

export class Filter extends Component {
    constructor(props) {
        super(props);

        this.state = {
            value: DEFAULT_STATE_DASHBOARDS_FILTER_NAME,
        };
    }

    componentWillReceiveProps(nextProps) {
        this.setState({
            value: nextProps.name,
        });
    }

    setFilterValue = event => {
        event.preventDefault();

        this.props.onChangeName(event.target.value);
    };

    handleKeyUp = event => {
        switch (event.keyCode) {
            case KEYCODE_ENTER:
                this.props.onKeypressEnter();
                break;
            case KEYCODE_ESCAPE:
                this.props.onChangeName();
                break;
            default:
                break;
        }
    };

    render() {
        const { classes, name, onChangeName } = this.props;

        const startAdornment = (
            <InputAdornment position="start">
                <SearchIcon className={classes.searchIcon} />
            </InputAdornment>
        );

        const endAdornment =
            name !== '' && name !== null ? (
                <InputAdornment position="end">
                    <ClearButton onClear={() => onChangeName()} />
                </InputAdornment>
            ) : null;

        return (
            <InputField
                className={classes.filterField}
                placeholder={i18n.t('Search for a dashboard')}
                startAdornment={startAdornment}
                endAdornment={endAdornment}
                value={this.state.value}
                onChange={this.setFilterValue}
                onKeyUp={this.handleKeyUp}
            />
        );
    }
}

Filter.propTypes = {
    name: PropTypes.string,
    onChangeName: PropTypes.func,
};

Filter.defaultProps = {
    name: '',
    onChangeName: Function.prototype,
};

export default withStyles(styles)(Filter);
