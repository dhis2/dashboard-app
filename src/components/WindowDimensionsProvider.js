import React, { createContext, useContext, useState, useEffect } from 'react'
import PropTypes from 'prop-types'
import debounce from 'lodash/debounce'

export const WindowDimensionsCtx = createContext(null)

const windowDims = () => ({
    height: window.innerHeight,
    width: window.innerWidth,
})

const WindowDimensionsProvider = ({ children }) => {
    const [dimensions, setDimensions] = useState(windowDims())

    useEffect(() => {
        const handleResize = debounce(() => {
            setDimensions(windowDims())
        }, 200)
        window.addEventListener('resize', handleResize)
        return () => {
            window.removeEventListener('resize', handleResize)
        }
    }, [])

    return (
        <WindowDimensionsCtx.Provider value={dimensions}>
            {children}
        </WindowDimensionsCtx.Provider>
    )
}

WindowDimensionsProvider.propTypes = {
    children: PropTypes.node.isRequired,
}

export default WindowDimensionsProvider

export const useWindowDimensions = () => {
    return useContext(WindowDimensionsCtx)
}
