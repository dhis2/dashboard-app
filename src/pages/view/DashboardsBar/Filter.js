import React, { useState } from 'react'
import { connect } from 'react-redux'
import PropTypes from 'prop-types'
import i18n from '@dhis2/d2-i18n'
import cx from 'classnames'
import { colors, IconSearch16, IconSearch24 } from '@dhis2/ui'

import ClearButton from './ClearButton'

import { useWindowDimensions } from '../../../components/WindowDimensionsProvider'
import {
    acSetDashboardsFilter,
    acClearDashboardsFilter,
} from '../../../actions/dashboardsFilter'
import { sGetDashboardsFilter } from '../../../reducers/dashboardsFilter'
import { isSmallScreen } from '../../../modules/smallScreen'

import classes from './styles/Filter.module.css'

export const KEYCODE_ENTER = 13
export const KEYCODE_ESCAPE = 27

const Filter = ({
    clearDashboardsFilter,
    expanded,
    filterText,
    setDashboardsFilter,
    onKeypressEnter,
    onSearchClicked,
}) => {
    const [focusedClassName, setFocusedClassName] = useState('')
    const [inputFocused, setInputFocus] = useState(false)
    const { width } = useWindowDimensions()

    const setFilterValue = event => {
        event.preventDefault()
        setDashboardsFilter(event.target.value)
    }

    const onKeyUp = event => {
        switch (event.keyCode) {
            case KEYCODE_ENTER:
                onKeypressEnter()
                clearDashboardsFilter()
                break
            case KEYCODE_ESCAPE:
                clearDashboardsFilter()
                break
            default:
                break
        }
    }

    const onFocus = event => {
        event.preventDefault()
        setFocusedClassName(classes.focused)
    }

    const onBlur = event => {
        event.preventDefault()
        setFocusedClassName('')
    }

    const onFocusInput = input => {
        if (input && inputFocused && isSmallScreen(width)) {
            return input.focus()
        }
    }

    const activateSearchInput = () => {
        onSearchClicked()
        setInputFocus(true)
    }

    return (
        <span
            className={cx(
                classes.container,
                expanded ? classes.expanded : classes.collapsed
            )}
        >
            <button
                className={classes.searchButton}
                onClick={activateSearchInput}
            >
                <IconSearch24 color={colors.grey800} />
            </button>
            <div
                className={cx(classes.searchArea, `${focusedClassName}`)}
                onFocus={onFocus}
                onBlur={onBlur}
            >
                <div className={classes.searchIconContainer}>
                    <IconSearch16 color={colors.grey800} />
                </div>
                <input
                    className={classes.input}
                    type="text"
                    placeholder={i18n.t('Search for a dashboard')}
                    onChange={setFilterValue}
                    onKeyUp={onKeyUp}
                    value={filterText}
                    data-test="search-dashboard-input"
                    ref={onFocusInput}
                />
                {filterText && (
                    <div className={classes.clearButtonContainer}>
                        <ClearButton onClear={clearDashboardsFilter} />
                    </div>
                )}
            </div>
        </span>
    )
}

Filter.propTypes = {
    clearDashboardsFilter: PropTypes.func,
    expanded: PropTypes.bool,
    filterText: PropTypes.string,
    setDashboardsFilter: PropTypes.func,
    onKeypressEnter: PropTypes.func,
    onSearchClicked: PropTypes.func,
}

const mapStateToProps = state => ({
    filterText: sGetDashboardsFilter(state),
})

const mapDispatchToProps = {
    setDashboardsFilter: acSetDashboardsFilter,
    clearDashboardsFilter: acClearDashboardsFilter,
}

export default connect(mapStateToProps, mapDispatchToProps)(Filter)
