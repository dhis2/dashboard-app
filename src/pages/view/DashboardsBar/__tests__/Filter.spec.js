import React from 'react'
import { mount } from 'enzyme'
import { render } from '@testing-library/react'
import { Provider } from 'react-redux'
import configureMockStore from 'redux-mock-store'
import Filter, {
    FilterUnconnected,
    KEYCODE_ENTER,
    KEYCODE_ESCAPE,
} from '../Filter'
import WindowDimensionsProvider from '../../../../components/WindowDimensionsProvider'

const mockStore = configureMockStore()

describe('Filter', () => {
    it('matches the snapshot when filterText property is empty', () => {
        const store = {
            dashboardsFilter: '',
        }
        const props = { classes: {} }
        const { container } = render(
            <Provider store={mockStore(store)}>
                <WindowDimensionsProvider>
                    <Filter {...props} />
                </WindowDimensionsProvider>
            </Provider>
        )
        expect(container).toMatchSnapshot()
    })

    describe('when filterText property is provided', () => {
        it('renders an input with correct value property', () => {
            const store = {
                dashboardsFilter: 'rainbow',
            }

            const props = { classes: {} }
            const { container } = render(
                <Provider store={mockStore(store)}>
                    <WindowDimensionsProvider>
                        <Filter {...props} />
                    </WindowDimensionsProvider>
                </Provider>
            )

            expect(container).toMatchSnapshot()
        })
    })

    describe('when input value is changed', () => {
        it('triggers setDashboardsFilter property with correct value', () => {
            const props = {
                filterText: '',
                setDashboardsFilter: jest.fn(),
                clearDashboardsFilter: jest.fn(),
                onKeypressEnter: jest.fn(),
                classes: {},
            }
            const newValue = 'fluttershy'
            const e = {
                target: { value: newValue },
                preventDefault: jest.fn(),
            }
            const filter = mount(
                <WindowDimensionsProvider>
                    <FilterUnconnected {...props} />
                </WindowDimensionsProvider>
            )
            filter.find('input').simulate('change', e)
            expect(e.preventDefault).toHaveBeenCalledTimes(1)
            expect(props.setDashboardsFilter).toHaveBeenCalledTimes(1)
            expect(props.setDashboardsFilter).toHaveBeenCalledWith(newValue)
        })
    })

    describe('when key is pressed', () => {
        it('triggers onKeypressEnter when key is ENTER', () => {
            const props = {
                filterText: '',
                clearDashboardsFilter: jest.fn(),
                onKeypressEnter: jest.fn(),
                classes: {},
            }
            const filter = mount(
                <WindowDimensionsProvider>
                    <FilterUnconnected {...props} />
                </WindowDimensionsProvider>
            )
            filter.find('input').simulate('keyUp', { keyCode: KEYCODE_ENTER })

            expect(props.onKeypressEnter).toHaveBeenCalledTimes(1)
            expect(props.clearDashboardsFilter).toHaveBeenCalledTimes(1)
        })

        it('triggers setDashboardsFilter when key is ESCAPE', () => {
            const props = {
                filterText: '',
                clearDashboardsFilter: jest.fn(),
                onKeypressEnter: jest.fn(),
                classes: {},
            }
            const filter = mount(
                <WindowDimensionsProvider>
                    <FilterUnconnected {...props} />
                </WindowDimensionsProvider>
            )

            filter.find('input').simulate('keyUp', { keyCode: KEYCODE_ESCAPE })

            expect(props.clearDashboardsFilter).toHaveBeenCalledTimes(1)
            expect(props.onKeypressEnter).not.toHaveBeenCalled()
        })
    })
})
