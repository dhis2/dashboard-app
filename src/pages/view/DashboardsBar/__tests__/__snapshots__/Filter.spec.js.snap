// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Filter matches the snapshot when filterText property is empty 1`] = `
<div>
  <div
    class="container collapsed"
  >
    <button
      class="searchButton"
    >
      <svg
        color="#404b5a"
        height="24"
        viewBox="0 0 24 24"
        width="24"
        xmlns="http://www.w3.org/2000/svg"
      >
        <path
          d="M10 3a7 7 0 015.6 11.201.056.056 0 01.013.009l.094.083 5 5a1 1 0 01-1.32 1.497l-.094-.083-5-5a1.009 1.009 0 01-.094-.108A7 7 0 1110 3zm0 2a5 5 0 100 10 5 5 0 000-10z"
          fill="currentColor"
        />
      </svg>
    </button>
    <div
      class="searchArea"
    >
      <div
        class="searchIconContainer"
      >
        <svg
          color="#404b5a"
          height="16"
          viewBox="0 0 16 16"
          width="16"
          xmlns="http://www.w3.org/2000/svg"
        >
          <path
            d="M6 1a5 5 0 013.871 8.164l4.483 4.482-.708.708L9.164 9.87A5 5 0 116 1zm0 1a4 4 0 100 8 4 4 0 000-8z"
            fill="currentColor"
          />
        </svg>
      </div>
      <input
        class="input"
        data-test="search-dashboard-input"
        placeholder="Search for a dashboard"
        type="text"
        value=""
      />
      
    </div>
  </div>
</div>
`;

exports[`Filter when filterText property is provided renders an input with correct value property 1`] = `
<div>
  <div
    class="container collapsed"
  >
    <button
      class="searchButton"
    >
      <svg
        color="#404b5a"
        height="24"
        viewBox="0 0 24 24"
        width="24"
        xmlns="http://www.w3.org/2000/svg"
      >
        <path
          d="M10 3a7 7 0 015.6 11.201.056.056 0 01.013.009l.094.083 5 5a1 1 0 01-1.32 1.497l-.094-.083-5-5a1.009 1.009 0 01-.094-.108A7 7 0 1110 3zm0 2a5 5 0 100 10 5 5 0 000-10z"
          fill="currentColor"
        />
      </svg>
    </button>
    <div
      class="searchArea"
    >
      <div
        class="searchIconContainer"
      >
        <svg
          color="#404b5a"
          height="16"
          viewBox="0 0 16 16"
          width="16"
          xmlns="http://www.w3.org/2000/svg"
        >
          <path
            d="M6 1a5 5 0 013.871 8.164l4.483 4.482-.708.708L9.164 9.87A5 5 0 116 1zm0 1a4 4 0 100 8 4 4 0 000-8z"
            fill="currentColor"
          />
        </svg>
      </div>
      <input
        class="input"
        data-test="search-dashboard-input"
        placeholder="Search for a dashboard"
        type="text"
        value="rainbow"
      />
      <div
        class="clearButtonContainer"
      >
        <button
          class="button"
        >
          <span>
            <svg
              class="icon"
              fill="#404b5a"
              viewBox="0 0 24 24"
              xmlns="http://www.w3.org/2000/svg"
            >
              <path
                d="M19 6.41L17.59 5 12 10.59 6.41 5 5 6.41 10.59 12 5 17.59 6.41 19 12 13.41 17.59 19 19 17.59 13.41 12z"
              />
              <path
                d="M0 0h24v24H0z"
                fill="none"
              />
            </svg>
          </span>
        </button>
      </div>
    </div>
  </div>
</div>
`;
